cpp/src/structs/utilities.cpp

```diff
@@ -168,15 +170,138 @@ flatten_nested_columns(table_view const& input,
+                                               cudf::column_view const& blueprint)
+{
+  // "Consume" columns from `flattened`, starting at `current_index`,
+  // based on the provided `blueprint` struct col. Recurse for struct children.
+  CUDF_EXPECTS(blueprint.type().id() == type_id::STRUCT,
+               "Expected blueprint column to be a STRUCT column.");
+
+  CUDF_EXPECTS(current_index < flattened.size(), "STRUCT column can't have 0 children.");
+
+  auto num_rows = flattened[current_index]->size();
+
+  // cudf::flatten_nested_columns() executes depth first, and serializes the struct null vector
+  // before the child/member columns.
+  // E.g. STRUCT_1< STRUCT_2< A, B >, C > is flattened to:
+  //      1. Null Vector for STRUCT_1
+  //      2. Null Vector for STRUCT_2
+  //      3. Member STRUCT_2::A
+  //      4. Member STRUCT_2::B
+  //      5. Member STRUCT_1::C
+  //
+  // Extract null-vector *before* child columns are constructed.
+  auto struct_null_column_contents = flattened[current_index++]->release();
+
+  auto struct_members = vector_of_columns{};
+  struct_members.reserve(blueprint.num_children());
+
+  std::transform(blueprint.child_begin(),
+                 blueprint.child_end(),
+                 std::back_inserter(struct_members),
+                 unflattener{flattened, current_index});
+
+  return cudf::make_structs_column(num_rows,
+                                   std::move(struct_members),
+                                   UNKNOWN_NULL_COUNT,  // Do count?
+                                   std::move(*struct_null_column_contents.null_mask));
+}
+
+bool is_struct(cudf::column_view const& col) { return col.type().id() == type_id::STRUCT; }
+
+/**
+ * @brief Check whether the specified column is of type LIST,
+ *        or any LISTs in its descendent columns.
+ *
+ */
+bool is_or_has_lists(cudf::column_view const& col)
+{
+  auto is_list = [](cudf::column_view const& col) { return col.type().id() == type_id::LIST; };
+
+  return is_list(col) ||
+         (is_struct(col) && std::any_of(col.child_begin(), col.child_end(), is_or_has_lists));
+}
+
+}  // namespace
+
+/**
+ * @copydoc cudf::structs::detail::unflatten_nested_columns
+ */
+std::unique_ptr<cudf::table> unflatten_nested_columns(std::unique_ptr<cudf::table>&& flattened,
+                                                      table_view const& blueprint)
+{
+  // Bail, if LISTs are present.
+  auto const has_lists = std::any_of(blueprint.begin(), blueprint.end(), is_or_has_lists);
+  CUDF_EXPECTS(not has_lists, "Unflattening LIST columns is not supported.");
+
+  // If there are no STRUCTs, unflattening is a NOOP.
+  auto const has_structs = std::any_of(blueprint.begin(), blueprint.end(), is_struct);
+  if (not has_structs) {
+    return std::move(flattened);  // Unchanged.
+  }
+
+  // There be struct columns.
+  // Note: Requires null vectors for all struct input columns.
+  // TODO: Explore if blueprint's struct's has_nulls() should be used at all.
```

Comment:
Not done yet?

Reply:
I can remove the `TODO`. I was hoping the reviewer might have a suggestion on whether the `blueprint` column should indicate if there are nulls in the `structs`. 
Line#291 indicates why this might not be possible. I'll remove the TODO.